File: solutions/typescript/bob/2/bob.ts
export function hey(message: string): string {
  message = message.trim();
  if (isSilence(message))
    return SILENCE_REPLY;
  if (isYelling(message)) {
    if (isQuestion(message))
      return YELLING_QUESTION_REPLY;
    else
      return YELLING_REPLY;
  }
  if (isQuestion(message))
    return QUESTION_REPLY;
  return DEFAULT_REPLY;
}

const SILENCE_REPLY = "Fine. Be that way!";
const YELLING_QUESTION_REPLY = "Calm down, I know what I'm doing!";
const YELLING_REPLY = "Whoa, chill out!";
const QUESTION_REPLY = "Sure.";
const DEFAULT_REPLY = "Whatever.";

function isYelling(message: string): boolean {
  return /[A-Z]+/.test(message) && message.toUpperCase() === message;
}

function isQuestion(message: string): boolean {
  return message.endsWith('?');
}

function isSilence(message: string): boolean {
  return message === "";
}

File: solutions/csharp/bob/2/Bob.cs
public static class Bob
{
    private static string QuestionResponse = "Sure.";
    private static string YellingResponse = "Whoa, chill out!";
    private static string YellingQuestionResponse = "Calm down, I know what I'm doing!";
    private static string SilenceResponse = "Fine. Be that way!";
    private static string DefaultResponse = "Whatever.";

    private static bool IsQuestion(string statement)
    {
        return statement.TrimEnd().EndsWith("?");
    }

    private static bool IsYelling(string statement)
    {
        bool hasLetters = false;
        foreach (char ch in statement)
        {
            if (char.IsLetter(ch))
            {
                if (!hasLetters)
                {
                    hasLetters = true;
                }
                if (char.IsLower(ch))
                {
                    return false;
                }
            }
        }
        return hasLetters;
    }

    private static bool IsSilence(string statement)
    {
        return string.IsNullOrWhiteSpace(statement);
    }

    public static string Response(string statement)
    {
        if (IsSilence(statement))
        {
            return SilenceResponse;
        }
        if (IsYelling(statement))
        {
            if (IsQuestion(statement))
            {
                return YellingQuestionResponse;
            }
            return YellingResponse;
        }
        if (IsQuestion(statement))
        {
            return QuestionResponse;
        }
        return DefaultResponse;
    }
}
